### PR TITLE
RFC-2841: add codegen flag export symbols from executable

### LINKED ISSUE #84161: Tracking Issue for `-C export-executable-symbols`
This is a tracking issue for the RFC "2841" (rust-lang/rfcs#2841).
The feature gate for the issue is `#![feature(export_executable_symbols)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [ ] Implement the RFC (cc @rust-lang/compiler  -- can anyone write up mentoring
      instructions?)
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

- Is this a good name for it?
- Should it be more general and export when limit_rdylib_exports or crate_type == ProcMacro?

### PR DESCRIPTION
**Pull request by csmoe**
Closes #84161
r? @nikomatsakis @Mark-Simulacrum

### CODE REVIEW

**Comment by crlf0710** [comment]
@csmoe Ping from triage, CI is still red here. Would you mind fixing that?

File: compiler/rustc_interface/src/tests.rs [hunk 1]
```diff
@@ -542,6 +542,7 @@ fn test_codegen_options_tracking_hash() {
     tracked!(debug_assertions, Some(true));
     tracked!(debuginfo, 0xdeadbeef);
     tracked!(embed_bitcode, false);
+    tracked!(export_executable_symbols, false);
```
Comment (bjorn3) [inline compiler/rustc_interface/src/tests.rs:545]:
```suggestion
    tracked!(export_executable_symbols, true);
```
the value needs to be different from the default value here.

File: compiler/rustc_session/src/options.rs [hunk 2]
```diff
@@ -891,6 +891,8 @@ options! {
         "allow the linker to link its default libraries (default: no)"),
     embed_bitcode: bool = (true, parse_bool, [TRACKED],
         "emit bitcode in rlibs (default: yes)"),
+    export_executable_symbols: bool =(false, parse_bool, [TRACKED],
```
Comment (bjorn3) [inline compiler/rustc_session/src/options.rs:894]:
```suggestion
    export_executable_symbols: bool = (false, parse_bool, [TRACKED],
```
formatting nit. rustfmt doesn't format macro invocations.

File: compiler/rustc_interface/src/tests.rs (hunk 1 above)
Comment (csmoe) [reply compiler/rustc_interface/src/tests.rs:545]: thanks you, I'd fix it soon

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Hmm, do we have any tests for symbol visibility? Maybe in the run-make folder? It'd be good to have a test for this.

**Comment by crlf0710** [comment]
@csmoe Ping from triage, would you mind adding the tests?

File: src/test/run-make/export-executable-symbols/main.rs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// edition:2018
+
+fn main() {
+    foo();
+}
+
+#[no_mangle]
+pub extern "C" fn foo() -> i32 {
```
Comment (nikomatsakis) [inline src/test/run-make/export-executable-symbols/main.rs:8]: Don't we need to write a `Makefile` that checks the flag has some kind of effect?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/test/run-make/export-executable-symbols/main.rs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+extern crate libc;
+use std::ffi::*;
+use std::os::unix::ffi::*;
+
+fn main() {
+    let path = std::env::var("TMPDIR").unwrap();
+    let path = std::path::PathBuf::from(path).join("libfoo.so");
+
+    let s = CString::new(path.as_os_str().as_bytes()).unwrap();
+    let handle = unsafe { libc::dlopen(s.as_ptr(), libc::RTLD_LAZY | libc::RTLD_LOCAL) };
```
Comment (bjorn3) [inline src/test/run-make/export-executable-symbols/main.rs:14]: I think this test will need to be restricted to unix targets or unix and windows targets if the windows api is used. Wasm can't support this either way.

Comment (nikomatsakis) [reply src/test/run-make/export-executable-symbols/main.rs:14]: Sounds reasonable but I'm not sure how to do that!

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/test/run-make/export-executable-symbols/main.rs (hunk 4 above)
Comment (nikomatsakis) [reply src/test/run-make/export-executable-symbols/main.rs:14]: Probably some annotation on the test file?

Comment (bjorn3) [reply src/test/run-make/export-executable-symbols/main.rs:14]: The `Makefile` accepts the same test flags as rust files in `test/ui`.

**Comment by csmoe** [comment]
@bjorn3  `-Cexport-executable-symbols` seems not work at all as `dlopen` errrors with `undefined symbol: exported_symbol`, `RUSTC_LOG=rustc_codegen_ssa::back::linker=debug` printed that the symbol was exported indeed:
```
 INFO rustc_codegen_ssa::back::link preparing Executable to "main"
DEBUG rustc_codegen_ssa::back::linker EXPORTED SYMBOLS:
DEBUG rustc_codegen_ssa::back::linker     exported_symbol;
DEBUG rustc_codegen_ssa::back::linker     main;
DEBUG rustc_codegen_ssa::back::linker     rust_eh_personality;
```
Need more help to move on :crying_cat_face:

**Comment by bjorn3** [comment]
Are you on Linux or macOS? What is the full linker invocation?

**Comment by csmoe** [comment]
```txt
"cc" "-Wl,--version-script=/repo/rust/./rustcqC1COd/list" "-m64" "main.main.cbc7ddee-cgu.0.rcgu.o" "main.main.cbc7ddee-cgu.1.rcgu.o" "main.main.cbc7ddee-cgu.10.rcgu.o" "main.main.cbc7ddee-cgu.11.rcgu.o" "main.main.cbc7ddee-cgu.12.rcgu.o" "main.main.cbc7ddee-cgu.13.rcgu.o" "main.main.cbc7ddee-cgu.14.rcgu.o" "main.main.cbc7ddee-cgu.15.rcgu.o" "main.main.cbc7ddee-cgu.2.rcgu.o" "main.main.cbc7ddee-cgu.3.rcgu.o" "main.main.cbc7ddee-cgu.4.rcgu.o" "main.main.cbc7ddee-cgu.5.rcgu.o" "main.main.cbc7ddee-cgu.6.rcgu.o" "main.main.cbc7ddee-cgu.7.rcgu.o" "main.main.cbc7ddee-cgu.8.rcgu.o" "main.main.cbc7ddee-cgu.9.rcgu.o" "main.4kqwyhdgeomtzpxx.rcgu.o" "-Wl,--as-needed" "-L" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-ea06e5b9ebe84ef4.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-608126472e4595bf.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-96b872de832ae456.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-c96f7126ed0a8243.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-d4333fc5a8e6acd5.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-286e391a7bef4a33.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-9b3b5b41ef9d89ac.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-a01624f3c4d2f24b.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-606d386967062702.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-9fec6174a0598006.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-612142634ae08506.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-8a18ecd4776dc12c.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-11234a29b766d417.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-5c49a6fde622d4a9.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-01242e2bb437fc18.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-a1d9b6c563f72c92.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-5a76a6f677cfd881.rlib" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-99c01d738b9e1246.rlib" "-Wl,--end-group" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-8ee0ab4b35b97256.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/repo/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "main" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs"
```
@bjorn3 linux, cc args as above.

**Comment by bjorn3** [comment]
Could you compile with `-Csave-temps` and post contents of the version script passed to `-Wl,--version-script=`?

**Comment by bjorn3** [comment]
Maybe `--gc-keep-exported` is necessar?

> When --gc-sections is enabled, this option prevents garbage collection of unused input sections that contain global symbols having default or protected visibility.  This
> option is intended to be used for executables where unreferenced sections would otherwise be garbage collected regardless of the external visibility of contained symbols.
> Note that this option has no effect when linking shared objects since it is already the default behaviour.  This option is only supported for ELF format targets.

**Comment by csmoe** [comment]
@bjorn3 version-script:
```txt
08:40 $ cat rustcXKEpHW/list
{
  global:
    exported_symbol;
    main;
    rust_eh_personality;

  local:
    *;
};
```
`--gc-keep-exported` is not available in old gcc(v6.3.0), I'll try a newer.

**Comment by csmoe** [comment]
gcc 10 with `--gc-keep-sections` doesn't help here.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@csmoe what is the status of this PR?

**Review by nikomatsakis — APPROVED** [review]
For my part, r+, but perhaps @bjorn3 knows of a blocker here?

**Comment by bjorn3** [comment]
Last time I checked the test didn't pass yet. Seems to have changed now. It looks like the test will only work on non-macOS unix targets as it hard-codes availability of `dlopen` (not available on windows) and the filename `libfoo.so` (macOS calls it `libfoo.dylib`).

File: compiler/rustc_session/src/options.rs [hunk 5]
```diff
@@ -1021,6 +1021,8 @@ options! {
         "allow the linker to link its default libraries (default: no)"),
     embed_bitcode: bool = (true, parse_bool, [TRACKED],
         "emit bitcode in rlibs (default: yes)"),
+    export_executable_symbols: bool = (false, parse_bool, [TRACKED],
```
Comment (bjorn3) [inline compiler/rustc_session/src/options.rs:1024]: Could you please move this from -C to -Z to prevent instantly stabilizing this?

**Review by bjorn3 — APPROVED** [review]
r=me with review comment fixed

**Comment by bjorn3** [comment]
There is a merge conflict.

**Comment by csmoe** [comment]
@bjorn3 conflicts resolved.

**Comment by csmoe** [comment]
```txt
= note: "arm-none-eabi-gcc" "-Wl,--version-script=/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/export-executable-symbols/export-executable-symbols/rustcDcYGyu/list" "-m64" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/export-executable-symbols/export-executable-symbols/libfoo.foo.91a024fd-cgu.0.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/export-executable-symbols/export-executable-symbols/libfoo.4ya8s8n1o2es9g07.rcgu.o" "-Wl,--as-needed" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/export-executable-symbols/export-executable-symbols" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-fcd8278774b9c6fc.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-e7f86684b9679284.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-10a87791239bd676.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-0e6cad6dd623d38e.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-9d05dd05a4d51b33.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-a60ff138fe1adfad.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-0e9344458c09713e.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-ef3b808d96f639bf.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-e3ce420901586b53.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-546bece8dd1bea42.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-9b358af16d7d7aa1.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-554cd499fe1a3b52.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-d105877dbe329cd1.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-8fec8a74c706e4b4.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-e869a247063ed69a.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-4896c4057dc9553a.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-2a6a2797f7a73818.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-0e3656b1fda5fd7b.rlib" "-Wl,--end-group" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-1f103368fa522bc0.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/export-executable-symbols/export-executable-symbols/libfoo.so" "-Wl,--gc-sections" "-shared" "-Wl,-zrelro,-znow" "-nodefaultlibs"
  = note: arm-none-eabi-gcc: error: unrecognized command line option '-m64'
```
have no idea where does the `m64` come from

**Comment by Dylan-DPC** [comment]
@csmoe any updates on this?

**Comment by Dylan-DPC** [comment]
@bjorn3 this is ready for review

**Comment by bjorn3** [comment]
I thought this was waiting on fixing a test. My bad.

@bors r+

**Comment by bjorn3** [comment]
Nope, the test wasn't fixed yet after all.

**Comment by csmoe** [comment]
> Nope, the test wasn't fixed yet after all.

will take a deep look soon

**Comment by csmoe** [comment]
I think it's fixed now. `-m64` came from missing `$TARGET` arg.

@bjorn3 r?

**Review by bjorn3 — APPROVED** [review]

**Comment by matthiaskrgr** [comment]
@bors r- rollup=iffy 
Failed in a rollup: https://github.com/rust-lang/rust/pull/98494#issuecomment-1166360623

**Comment by csmoe** [comment]
Ignore windows and no-std targets as the test depends on `dlopen` and `std`.
@bjorn3 r?

**Comment by bjorn3** [comment]
I think you will also need to ignore wasm.

**Comment by csmoe** [comment]
CI setting up failed.

在 2022年7月1日，下午12:11，Rust Log Analyzer ***@***.***> 写道：

﻿

A job failed! Check out the build log: (web)<https://github.com/rust-lang-ci/rust/runs/7143639465?check_suite_focus=true> (plain)<https://github.com/rust-lang-ci/rust/commit/20f992599c2cd217ac8b6bbabc8aa942c4abc11d/checks/7143639465/logs>

Click to see the possible cause of the failure (guessed by this bot)

**Comment by csmoe** [comment]
Is there anyway to run all the test including all the targets locally? wasm seems not excluded.
(this pr takes so long, I wanna make sure it works before next rollup)

**Comment by bjorn3** [comment]
Right, this needs to be updated for https://github.com/rust-lang/rust/pull/98975.

**Comment by csmoe** [comment]
@bjorn3 I found the root cause of missing `exported_symbol` in the final executable: the existing linker arg `-Wl,--gc-sections` strip the symbol. Should I overwrite this link arg if `exported-executable-symbol=yes`? Or warning the user to add `-Wl,--no-gc-sectioins` manually?

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 6]
```diff
@@ -2082,7 +2082,12 @@ fn add_order_independent_options(
         // sections to ensure we have all the data for PGO.
         let keep_metadata =
             crate_type == CrateType::Dylib || sess.opts.cg.profile_generate.enabled();
-        cmd.gc_sections(keep_metadata);
+
+        if !(crate_type == CrateType::Executable
+            && sess.opts.unstable_opts.export_executable_symbols)
+        {
+            cmd.gc_sections(keep_metadata);
+        }
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/link.rs:2090]: Maybe move the `no_gc_sections` call here if `export_executable_symbols` is true.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 7]
```diff
@@ -2082,7 +2082,12 @@ fn add_order_independent_options(
         // sections to ensure we have all the data for PGO.
         let keep_metadata =
             crate_type == CrateType::Dylib || sess.opts.cg.profile_generate.enabled();
-        cmd.gc_sections(keep_metadata);
+
+        if !(crate_type == CrateType::Executable
+            && sess.opts.unstable_opts.export_executable_symbols)
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/link.rs:2087]: Maybe use `crate_type != Executable || !export_executable_symbols`?

**Comment by bjorn3** [comment]
I am not entirely confident in `-Zexport-executable-symbols` being implemented correctly here, but I am confident that this doesn't break anything if `-Zexport-executable-symbols` isn't used. As such r=me with or without my nits above.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 8]
```diff
@@ -2082,7 +2082,12 @@ fn add_order_independent_options(
         // sections to ensure we have all the data for PGO.
         let keep_metadata =
             crate_type == CrateType::Dylib || sess.opts.cg.profile_generate.enabled();
-        cmd.gc_sections(keep_metadata);
+        if crate_type == CrateType::Executable && !sess.opts.unstable_opts.export_executable_symbols
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/link.rs:2085]:
```suggestion
        if crate_type != CrateType::Executable || !sess.opts.unstable_opts.export_executable_symbols
```

**Comment by bjorn3** [comment]
The test needs to be ignored on nvptx64-nvidia-cuda. Maybe an `// only-unix` annotation works? Not sure if that is valid though.

**Comment by csmoe** [comment]
`only-unix` seems not working, as it ignored `linux`. I just limited the test with `ignore-cross-compile`.

**Comment by bjorn3** [comment]
:+1: Could you please squash? I noticed that one of the commits introduced merge conflict markers which another solves again. And besides, this is pretty much an atimic change. r=me after squashing

**Comment by bjorn3** [comment]
That only took 10 tries...

**Pull request merged**